Gateway/Validator/CheckoutResponseValidator.php

```diff
@@ -95,23 +95,59 @@ public function validate(array $validationSubject)
 					}
 					break;
 				case "RedirectShopper":
-					$payment->setAdditionalInformation('3dActive', true);
-
-					if (!empty($paReq = $response['redirect']['data']['PaReq']) &&
-						!empty($md = $response['redirect']['data']['MD']) &&
-						!empty($issuerUrl = $response['redirect']['url']) &&
-						!empty($paymentData = $response['paymentData'])
-					) {
-						$payment->setAdditionalInformation('issuerUrl', $issuerUrl);
-						$payment->setAdditionalInformation('paRequest', $paReq);
-						$payment->setAdditionalInformation('md', $md);
-						$payment->setAdditionalInformation('paymentData', $paymentData);
+
+					$redirectUrl = null;
+					$paymentData = null;
+
+					if (isset($response['redirect']['url']) && !empty($response['redirect']['url'])) {
+						$redirectUrl = $response['redirect']['url'];
+					}
+
+					if (!empty($response['paymentData'])) {
+						$paymentData = $response['paymentData'];
+					}
+
+					// If the redirect data is there then the payment is a card payment with 3d secure
+					if (isset($response['redirect']['data'])) {
+						$paReq = null;
+						$md = null;
+
+						$payment->setAdditionalInformation('3dActive', true);
+
+						if (!empty($response['redirect']['data']['PaReq'])) {
+							$paReq = $response['redirect']['data']['PaReq'];
+						}
+
+						if (!empty($response['redirect']['data']['MD'])) {
+							$md = $response['redirect']['data']['MD'];
+						}
+
+						if ($paReq && $md && $redirectUrl && $paymentData) {
+							$payment->setAdditionalInformation('issuerUrl', $redirectUrl);
+							$payment->setAdditionalInformation('paRequest', $paReq);
+							$payment->setAdditionalInformation('md', $md);
+							$payment->setAdditionalInformation('paymentData', $paymentData);
+						} else {
+							$isValid = false;
+							$errorMsg = __('3D secure is not valid.');
+							$this->adyenLogger->error($errorMsg);;
+							$errorMessages[] = $errorMsg;
+						}
+					// otherwise it is an alternative payment method which only requires the redirect url to be present
 					} else {
-						$isValid = false;
-						$errorMsg = __('3D secure is not valid.');
-						$this->adyenLogger->error($errorMsg);;
-						$errorMessages[] = $errorMsg;
+						$payment->setAdditionalInformation('CheckoutAPM', true);
+
+						if ($redirectUrl && $paymentData) {
+							$payment->setAdditionalInformation('redirectUrl', $redirectUrl);
+							$payment->setAdditionalInformation('paymentData', $paymentData);
```

Comment:
do we clear paymentData after usage ?

Reply:
Should I do it with the paReq and MD as well?